### ISSUE TITLE
Retrieve a player's language using protocollib

### ISSUE THREAD

**Issue by KieronWiltshire** [BlackyPaw/I18N#2]
I created a custom LocaleResolver which retrieve's the player's language preference from the client. I think it'd be real neat to add this into the plugin by default, obviously you'd need to tweek it.

``` Java
package com.dracade.vone.api.utilities;

import com.blackypaw.mc.i18n.LocaleResolver;
import com.comphenix.protocol.PacketType;
import com.comphenix.protocol.events.ListeningWhitelist;
import com.comphenix.protocol.events.PacketEvent;
import com.comphenix.protocol.events.PacketListener;
import com.comphenix.protocol.injector.GamePhase;
import com.dracade.vone.api.Vone;
import org.bukkit.entity.Player;

import java.util.Locale;

public class PlayerLocaleResolver implements PacketListener, LocaleResolver {

    private Vone vone;

    /**
     * Create a new PlayerLocaleResolver instance.
     *
     * @param vone
     */
    public PlayerLocaleResolver(Vone vone) {
        this.vone = vone;
    }

    @Override
    public void onPacketSending(PacketEvent event) {
        return;
    }

    @Override
    public void onPacketReceiving(PacketEvent event) {
        if (!(event.getPacket().getType() == PacketType.Play.Client.SETTINGS)) return;

        if (event.getPlayer() != null) {
            String language = event.getPacket().getStrings().readSafely(0);
            this.trySetPlayerLocale(event.getPlayer(), new Locale(language));
        }
    }

    @Override
    public ListeningWhitelist getSendingWhitelist() {
        return ListeningWhitelist.newBuilder().gamePhase(GamePhase.PLAYING).highest().types(PacketType.Play.Client.SETTINGS).build();
    }

    @Override
    public ListeningWhitelist getReceivingWhitelist() {
        return ListeningWhitelist.newBuilder().gamePhase(GamePhase.PLAYING).highest().types(PacketType.Play.Client.SETTINGS).build();
    }

    @Override
    public Vone getPlugin() {
        return this.vone;
    }

    @Override
    public Locale resolveLocale(Player player) {
        return this.vone.getPlayerLocale(player);
    }

    @Override
    public boolean trySetPlayerLocale(Player player, Locale locale) {
        this.vone.setPlayerLocale(player, locale);
        return true;
    }

}


```

If the player's language isn't supported, then use the fallback language.

**Comment by cieldeville** [comment]
Dear Kieron,

I actually thought about adding this into the plugin when I first implemented the respective event. The main reason I decided against doing so was that the respective client options packet will be sent after Bukkit's login event gets dispatched yet I18N guarantees that players' locales are available as soon as all LoginEvent handlers with ListenerPriority.LOWEST have been invoked. Thus there would be a short time frame in which retrieving a player's locale would either return a possibly incorrect result (e.g. returning the fallback language) or nothing at all.

**Comment by KieronWiltshire** [comment]
Ah, so this piece of code is potentially pointless? I've yet to test it with multiple language files :/

**Comment by cieldeville** [comment]
It's not potentially pointless: the only problem you will have to consider is what you will return for players who have not yet sent the client options packet but who are supposed to be displayed a translated message. As I said above you may choose from two options: either return the fallback locale (what may result in messages being displayed in two different languages shortly after a player joined, e.g. join-message in English and all what follows in German) or you make this a special case for example by returning null. That would break I18N's promise, though, and will prevent you from using third-party plugins that might make good use of that promise. Also, doing so would require utmost care from yourself as you would have to constantly safe-guard from that special case.

TL;DR -> If you still want to use your locale resolver I would go with returning the default language. Otherwise I would rather make use of another locale resolver to get the last locale a player has had set for himself and listen to I18N's event that contains the language the player set for himself once that information is available. Afterwards you can silently adjust the player's locale to its new value in case it has changed. Also, as most players won't change their language too frequently this would result in a player's language being available for join-messages,too, as long as he did not change it.
